example/my_app.py

```diff
@@ -0,0 +1,106 @@
+        to execute code on SLURM cluster to test wandb pre-emption:
+          `python my_app.py --multirun +sweeper=wandb +launcher=submitit_remote`
+          then execute `scancel --signal=USR1 <JOB_ID>`
+
+          --> requires submit it launcher to be installed
+              `pip install hydra-submitit-launcher --upgrade`
+        ------------------------------------------------------------------------
+        """
+        dropout = cfg.model.dropout
+        batch_size = cfg.experiment.batch_size
+        database = cfg.experiment.db
+        lr = cfg.experiment.lr
+
+        # Here goes some long calculation
+        best = 1e8
+        num_iters = 100.0
+        for i in range(int(num_iters)):
+            self.ready_to_checkpoint = False
+
+            # NOTE: Be careful not to log too quickly, it could cause an os._exit(-1) from wandb which can't
+            # be caught for graceful exit. Since this task function is running on a thread spun by the agent,
+            # once an os._exit(-1) occurs from wandb, all parent threads are eliminated, including the launcher.
+            sleep(0.5)
+
+            noise = random.random() / 10.0
+            out = (
+                float(
+                    abs(dropout - 0.33)
+                    + int(database == "mnist")
+                    + abs(lr - 0.12)
+                    + abs(batch_size - 4)
+                )
+                + noise
+                + (2.0 ** ((-0.4 * i) + 3))
+            )
+            out = max(out, 0.0)
+            wandb.log({"out": out})
+            if (i + 1) % 10 == 0:
+                logger.info(f"{100*i/num_iters:.1f}% done...")
+            if out < best:
+                best = out
+
+            # Simulate a bunch of important computations needed before saving whatever
+            # you want to save, like model parameters.
+            # NOTE: Optional. Uncomment this if you want to further test out pre-emption.
+            # sleep(30)
+
+            # This is to signal that when pre-emption happens, wait until all important computations
+            # are finished before finishing execution of checkpoint function.
+            self.ready_to_checkpoint = True
+
+        logger.info(
+            f"best dummy_training(dropout={dropout:.3f}, lr={lr:.3f}, db={database}, batch_size={batch_size}) "
+            f"= {best:.3f}"
+        )
+
+        if cfg.error:
+            raise RuntimeError("cfg.error is True")
+
+        if cfg.return_type == "float":
+            return best
+
+        if cfg.return_type == "dict":
+            return dict(name="objective", type="objective", value=best)
+
+        if cfg.return_type == "list":
+            return [dict(name="objective", type="objective", value=best)]
+
+        if cfg.return_type == "none":
+            return None
+
+
+if __name__ == "__main__":
+    dummy_training = DummyTraining()
+    app = hydra.main(config_path="conf", config_name="config")(dummy_training.__call__)
```

Comment:
If this is a SLURM specific example, please mention that. I'd also request you to add a minimal example with local multiprocessing (with ray or joblib). That way, people who don't want to use SLURM can have something to refer to.

An example that can be invoked in the following way.
```bash
$ python3 main.py hydra/sweeper=wandb hydra/launcher=ray
$ python3 main.py hydra/sweeper=wandb hydra/launcher=joblib
```

Reply:
I mention it here https://github.com/captain-pool/hydra-wandb-sweeper/blob/e791aec734530bc9b611f938ca6a5962d67eb6e6/example/my_app.py#L19-L37

Submitit also supports local multiprocessing by switching to: `- override /hydra/launcher: submitit_local`
Documentation available [here](https://github.com/facebookincubator/submitit/blob/main/docs/examples.md) and [here](https://hydra.cc/docs/plugins/submitit_launcher/).

Explicit support for the other launchers, such as Ray and Joblib, go beyond the scope of this PR. If people want to do local multiprocessing they can use the submitit launcher but if they want to try out Ray or Joblib, that'll have to wait (I don't have the time since I made this PR for the sake of helping out my research and I need to get back to my research).

Reply:
Cool! Once this PR gets merged, I'll take look into ray / joblib. :) I had no idea about `submitit_local`. Thanks for the heads up.

Reply:
Just a heads up that this sweeper should be launcher agnostic as much as possible! So if it doesn't work for the other launchers lmk and we can figure out a way to future proof this. It would be a nightmare if down the road the other launchers get modified and suddenly this sweeper needs to be modified to reflect that change.